### PR TITLE
Add usernetes with docker example

### PR DESCRIPTION
**Pull request by afbjorklund**
Will run Kubernetes in rootless docker, a.k.a. "usernetes".

* https://github.com/rootless-containers/usernetes

* https://rootlesscontaine.rs/getting-started/docker

Add certificate for localhost so we can use it from the host.

Note: This is running kubernetes-in-docker ([kind](https://kind.sigs.k8s.io/))

Currently: `FROM docker.io/kindest/node:v1.28.0`

```
NAME           STATUS   ROLES           AGE   VERSION   INTERNAL-IP    EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION      CONTAINER-RUNTIME
u7s-lima-u7s   Ready    control-plane   15m   v1.28.0   192.168.5.15   <none>        Debian GNU/Linux 11 (bullseye)   5.15.0-86-generic   containerd://1.7.1
```

----
```
CONTAINER ID   IMAGE            COMMAND                  CREATED          STATUS          PORTS                                                                                                                                                                            NAMES
a666a9b648cb   usernetes-node   "/u7s-entrypoint.sh …"   20 minutes ago   Up 15 minutes   0.0.0.0:2379->2379/tcp, :::2379->2379/tcp, 0.0.0.0:6443->6443/tcp, :::6443->6443/tcp, 0.0.0.0:10250->10250/tcp, :::10250->10250/tcp, 0.0.0.0:8472->8472/udp, :::8472->8472/udp   usernetes-node-1
```
```
REPOSITORY       TAG                      IMAGE ID       CREATED          SIZE
usernetes-node   latest                   f5a3eccc00a4   22 minutes ago   1.09GB
kindest/node     v1.28.0                  ad70201dab13   3 months ago     950MB
debian           bullseye-20230814-slim   fb4ec5ceea2f   3 months ago     80.5MB
busybox          latest                   a416a98b71e2   4 months ago     4.26MB
```

### CODE REVIEW

**Comment by afbjorklund** [comment]
It is possible to run this example with podman, by changing some variables.

`export CONTAINER_ENGINE=podman`

File: examples/experimental/u7s.yaml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    curl -L "https://dl.k8s.io/release/$version/bin/linux/$arch/kubectl" -o /usr/local/bin/kubectl
+    chmod 755 /usr/local/bin/kubectl
+    kubectl version --client
+- mode: system
+  script: |
+    #!/bin/bash
+    set -eux -o pipefail
+    test -d /var/lib/usernetes && exit 0
+    branch=master
+    curl -fsSL https://github.com/rootless-containers/usernetes/archive/$branch.tar.gz | tar -C /var/lib --transform="s/-$branch//" -xz
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:50]: Why not git clone?

File: examples/experimental/u7s.yaml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    chown ${LIMA_CIDATA_UID} /var/lib/usernetes
+    cd /var/lib/usernetes
+    patch --forward -r - kubeadm-config.yaml <<EOF
+    @@ -7,6 +7,9 @@
+     ---
+     apiVersion: kubeadm.k8s.io/v1beta3
+     kind: ClusterConfiguration
+    +apiServer:
+    +  certSANs:
+    +  - "127.0.0.1"
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:60]: Can we us yq?

File: examples/experimental/u7s.yaml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    set -eux -o pipefail
+    cd /var/lib/usernetes
+    make up
+    make kubeadm-init
+    # Installing a Pod network add-on
+    make install-flannel
+    # Control plane node isolation
+    kubectl taint nodes --all node-role.kubernetes.io/control-plane-
+    make kubeconfig
+    # Replace the server address with localhost, so that it works also from the host
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:87]: Does it still work from the guest too?

File: examples/experimental/u7s.yaml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    esac
+    curl -L "https://dl.k8s.io/release/$version/bin/linux/$arch/kubectl" -o /usr/local/bin/kubectl
+    chmod 755 /usr/local/bin/kubectl
+    kubectl version --client
+- mode: system
+  script: |
+    #!/bin/bash
+    set -eux -o pipefail
+    test -d /var/lib/usernetes && exit 0
+    branch=master
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:49]: Can we pin the tag?

File: examples/experimental/u7s.yaml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  arch: "x86_64"
+- location: "https://cloud-images.ubuntu.com/releases/22.04/release/ubuntu-22.04-server-cloudimg-arm64.img"
+  arch: "aarch64"
+
+# Mounts are disabled in this template, but can be enabled optionally.
+mounts: []
+# containerd is managed by Docker, not by Lima, so the values are set to false here.
+containerd:
+  system: false
+  user: false
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:24]: Why not use nerdctl 🤓

File: examples/README.md [hunk 6]
```diff
@@ -45,6 +45,7 @@ Container orchestration:
 - [`faasd`](./faasd.yaml): [Faasd](https://docs.openfaas.com/deployment/faasd/)
 - [`k3s`](./k3s.yaml): Kubernetes via k3s
 - [`k8s`](./k8s.yaml): Kubernetes via kubeadm
+- [`experimental/u7s`](./experimental/u7s.yaml): Kubernetes (rootless) = "Usernetes"
```
Comment (AkihiroSuda) [inline examples/README.md:48]:
```suggestion
- [`experimental/u7s`](./experimental/u7s.yaml): [Usernetes](https://github.com/rootless-containers/usernetes): Rootless Kubernetes
```

**Comment by vsoch** [comment]
@afbjorklund how are you handling creating the join-command here (and sharing with a worker?) I'm trying to get that working for my setup, and it seems like the mounts are not done until the provision is done (so I can't write content there) and there doesn't seem to be a copy directive for into the build. I found limactl copy but that assumes both are running (and I'd like the second to provision with the file) https://fig.io/manual/limactl and I found this issue https://github.com/lima-vm/lima/issues/594 but it wasn't clear that there was a resolution. Thanks for the tips!

**Comment by jandubois** [comment]
> it seems like the mounts are not done until the provision is done

This is true for `reverse-sshfs`, but not for `9p` or `virtiofs`.

**Comment by vsoch** [comment]
> This is true for reverse-sshfs, but not for 9p or virtiofs.

Is there an example that shows how to do this? And so I understand, the reverse sshfs is like a network filesystem (using ssh) and then the others are different kinds of filesystems? I found:

- https://virtio-fs.gitlab.io/
- is this the right 9p? https://en.wikipedia.org/wiki/9P_(protocol) it's a hard one to search for! :laughing:

**Comment by jandubois** [comment]
https://github.com/lima-vm/lima/blob/a21b5f3bbcb63a37987a328e63d8a1a9f1c2e098/examples/default.yaml#L101-L104

File: examples/experimental/u7s.yaml (hunk 5 above)
Comment (afbjorklund) [reply examples/experimental/u7s.yaml:24]: Docker was recommended

**Comment by afbjorklund** [comment]
> how are you handling creating the join-command here (and sharing with a worker?)

I did not, it was something that was left-do-do also for the regular rootful Kubernetes...

* https://github.com/lima-vm/lima/issues/1184

File: examples/experimental/u7s.yaml (hunk 3 above)
Comment (afbjorklund) [reply examples/experimental/u7s.yaml:87]: It is binding to `0.0.0.0` by default

Comment (afbjorklund) [reply examples/experimental/u7s.yaml:87]: So that means "yes", from both.

```console
anders@ubuntu:~$ limactl shell u7s kubectl version
Client Version: v1.28.4
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.4
anders@ubuntu:~$ LIMA_INSTANCE=u7s kubectl.lima version
Client Version: v1.28.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.4

```

Comment (afbjorklund) [reply examples/experimental/u7s.yaml:87]: Same as with the regular k8s

File: examples/experimental/u7s.yaml (hunk 1 above)
Comment (afbjorklund) [reply examples/experimental/u7s.yaml:50]: More of a developer tool?

File: examples/experimental/u7s.yaml (hunk 2 above)
Comment (afbjorklund) [reply examples/experimental/u7s.yaml:60]: Need to install it.

```console
$ yq
Command 'yq' not found, but can be installed with:
snap install yq
Please ask your administrator.
```

**Comment by afbjorklund** [comment]
> * is this the right 9p?

https://wiki.qemu.org/Documentation/9psetup

You can also search for "VirtFS", the qemu setup:
[VirtFS—A virtualization aware File System pass-through](https://landley.net/kdocs/ols/2010/ols2010-pages-109-120.pdf) [PDF]

But yeah, it is the same old Plan 9 transport...

**Comment by afbjorklund** [comment]
Note: we scoped out the regular kind and k3d templates, and left it as an exercise for the reader

So we _could_ do the same with usernetes, and just refer to "docker.yaml" (rootless) as the base?

File: examples/experimental/u7s.yaml (hunk 5 above)
Comment (AkihiroSuda) [reply examples/experimental/u7s.yaml:24]: Nerdctl now achieves good compatibility with Docker, so we can recommend nerdctl too.

Comment (afbjorklund) [reply examples/experimental/u7s.yaml:24]: The `nerdctl compose` command failed with an error: `FATA[0000] getwd: no such file or directory`

Comment (afbjorklund) [reply examples/experimental/u7s.yaml:24]: Added an explicit CONTAINER_ENGINE variable, so that it is easy to test with the alternatives.

File: examples/experimental/u7s.yaml (hunk 4 above)
Comment (afbjorklund) [reply examples/experimental/u7s.yaml]: Pinned to `gen2-v20230919.0`

File: examples/experimental/u7s.yaml [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    curl -L "https://dl.k8s.io/release/$version/bin/linux/$arch/kubectl" -o /usr/local/bin/kubectl
+    chmod 755 /usr/local/bin/kubectl
+    kubectl version --client
+- mode: system
+  script: |
+    #!/bin/bash
+    set -eux -o pipefail
+    test -d /var/lib/usernetes && exit 0
+    branch=gen2-v20230919.0
+    kubernetes_version=1.28.0
```
Comment (afbjorklund) [inline examples/experimental/u7s.yaml:50]: Maybe get this from the code

```console
$ grep kindest/node Dockerfile
ARG BASE_IMAGE=docker.io/kindest/node:v1.28.0

```

Comment (afbjorklund) [reply examples/experimental/u7s.yaml:50]: It is not the same images, but close enough (same version)

**Comment by vsoch** [comment]
For those following the thread about virtio-fs - I found that the C variant that shipped with QEMU did not work. So I built the rust one, e.g.,

```bash
# This is in the PWD
git clone https://gitlab.com/virtio-fs/virtiofsd 
cd virtiofsd 
sudo apt install libcap-ng-dev libseccomp-dev
```

Then build with cargo.

```bash
cargo build --release
```

Then I replaced it.

```
sudo mv /usr/lib/qemu/virtiofsd /usr/lib/qemu/virtiofsd-c
sudo mv virtiofsd/target/release/virtiofsd /usr/lib/qemu/virtiofsd
```

I also did:

```
sudo usermod -aG kvm $USER
```

And then I was able to copy the join-command (or whatever files I needed) into the mount directory and they will be present when you are provisioning!

```bash
mkdir -p /tmp/lima
cp /home/vanessa/.lima/control-plane/join-command /tmp/lima/join-command
```

@afbjorklund @jandubois I removed my home `~` for this to work, but then it shells in and there is no such file or directory. Is there a way to change the present working directory? I already checked default.yaml and searched around for terms but didn't find anything. ty!

**Comment by jandubois** [comment]
> I removed my home `~` for this to work, but then it shells in and there is no such file or directory. Is there a way to change the present working directory?

I'm not sure what you mean by "it shells in". In general you can specify the working directory with `limactl shell $INSTANCE --workdir $WORKDIR ...` and for the `lima` wrapper you can set the default with `$LIMA_WORKDIR`.

But I'm not sure if this applies to your situation as I'm not clear about what "it" means...

**Comment by vsoch** [comment]
This works great!

```bash
limactl shell --workdir /opt/usernetes usernetes-worker
```
that should work for now. Thank you!

**Comment by vsoch** [comment]
Thanks for the help! I now have a worker and control plane set of configs, and it's entirely automated except for copying the join-command to the (TBA worker) directory that is mounted and available for provision, and then for actually running join I have the user do that interactively because otherwise there is an issue with containerd. But overall it's just a few commands and very easy, and that's great! Thanks for all the help! https://github.com/converged-computing/usernetes-lima

**Comment by afbjorklund** [comment]
Added a very important `sleep 5` (for containerd to start), and moved `make kubeconfig` above kubectl (doh!)

It would be nice if usernetes used something like `/etc/usernetes` for the kubeconfig, but that's another issue.

File: examples/experimental/u7s.yaml [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+provision:
+# See [KEP-2033](https://features.k8s.io/2033) Kubelet-in-UserNS (aka Rootless mode)
+- mode: system
+  script: |
+    #!/bin/bash
+    set -eux -o pipefail
+    test -d /var/lib/usernetes && exit 0
+    branch=gen2-v20230919.0
+    curl -fsSL https://github.com/rootless-containers/usernetes/archive/$branch.tar.gz | tar -C /var/lib --transform="s/-$branch//" -xz
+    chown ${LIMA_CIDATA_UID} /var/lib/usernetes
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:44]: `usernetes` is expected to be cloned in the (guest) home

File: examples/experimental/u7s.yaml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+  system: false
+  user: false
+provision:
+# See [KEP-2033](https://features.k8s.io/2033) Kubelet-in-UserNS (aka Rootless mode)
+- mode: system
+  script: |
+    #!/bin/bash
+    set -eux -o pipefail
+    test -d /var/lib/usernetes && exit 0
+    branch=gen2-v20230919.0
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:42]: tag, not branch

File: examples/experimental/u7s.yaml [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+  arch: "aarch64"
+
+# Mounts are disabled in this template, but can be enabled optionally.
+mounts: []
+# containerd is managed by Docker, not by Lima, so the values are set to false here.
+containerd:
+  system: false
+  user: false
+provision:
+# See [KEP-2033](https://features.k8s.io/2033) Kubelet-in-UserNS (aka Rootless mode)
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:36]: Doesn't seem relevant to the provision scripts

File: examples/experimental/u7s.yaml [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    esac
+    curl -L "https://dl.k8s.io/release/$version/bin/linux/$arch/kubectl" -o /usr/local/bin/kubectl
+    chmod 755 /usr/local/bin/kubectl
+    kubectl version --client
+- mode: system
+  script: |
+    #!/bin/bash
+    cd /var/lib/usernetes/init-host
+    export CONTAINER_ENGINE=docker
+    ./init-host.root.sh
```
Comment (AkihiroSuda) [inline examples/experimental/u7s.yaml:63]: Could you add a comment to explain that Docker is installed here?

File: examples/experimental/u7s.yaml (hunk 9 above)
Comment (afbjorklund) [reply examples/experimental/u7s.yaml:42]: I suppose it is the same parameter, especially if we use git instead.

File: examples/experimental/u7s.yaml (hunk 11 above)
Comment (afbjorklund) [reply examples/experimental/u7s.yaml:63]: Could just remove it, and go with whatever default usernetes has

File: examples/experimental/u7s.yaml (hunk 9 above)
Comment (AkihiroSuda) [reply examples/experimental/u7s.yaml:42]: Yes, but the variable is just a misnomer

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Comment by afbjorklund** [comment]
To access the control-plane (which is normally not needed, but whatever) one can do:

`limactl shell --workdir usernetes u7s make shell`

And like it says in the README, you need to edit the compose.yaml in order to add node ports.

But when you do, they are also forwarded by Lima...

**Review by AkihiroSuda — APPROVED** [review]
Thanks
